### PR TITLE
PUBDEV-4992: Incremental scoring on validation set + re-using predictions for GainsLift

### PR DESCRIPTION
**Pull request by michalkurka**

### CODE REVIEW

**Comment by michalkurka** [comment]
@h2o-ops please test!

**Comment by michalkurka** [comment]
@h2o-ops please test!

**Comment by michalkurka** [comment]
@h2o-ops please test!

**Comment by michalkurka** [comment]
@h2o-ops please test!

**Comment by michalkurka** [comment]
@h2o-ops please test!

**Comment by michalkurka** [comment]
@h2o-ops please test!

**Comment by michalkurka** [comment]
@h2o-ops please test!

**Comment by michalkurka** [comment]
@h2o-ops please test!

**Comment by michalkurka** [comment]
@h2o-ops please test!

**Comment by michalkurka** [comment]
Rebased to latest master

**Comment by michalkurka** [comment]
@h2o-ops please test!

**Comment by michalkurka** [comment]
@h2o-ops please test!

File: h2o-algos/src/main/java/hex/tree/Score.java [hunk 1]
```diff
@@ -17,19 +19,33 @@
   final boolean _oob;           // Computed on OOB
   final Key<Vec> _kresp;        // Response vector key (might be either train or validation)
   final ModelCategory _mcat;    // Model category (Binomial, Regression, etc)
-  ModelMetrics.MetricBuilder _mb;
-//  GainsLift.GainsLiftBuilder _gainsLiftBuilder;
+  ModelMetricsSupervised.MetricBuilderSupervised _mb;
   final boolean _computeGainsLift;
+  final ScoreIncInfo _sii;      // Incremental scoring (on a validation dataset), null indicates full scoring
+  final Frame _preds;           // Prediction cache (typically not too many Vecs => it is not too costly embed the object in MRTask)
```
Comment (mmalohlava) [inline h2o-algos/src/main/java/hex/tree/Score.java]: should be Key, no?

File: h2o-algos/src/main/java/hex/tree/Score.java [hunk 2]
```diff
@@ -17,19 +19,33 @@
...
+    _preds = computeGainsLift ? preds : null; // don't keep the prediction cache if we don't need to compute gainslift
+    assert (! _is_train) || (_sii == null);
+  }
 
-  @Override public void map( Chunk chks[] ) {
+  @Override public void map(Chunk allchks[]) {
+    final Chunk[] chks = getScoringChunks(allchks);
     Chunk ys = _bldr.chk_resp(chks);  // Response
-    Model m = _bldr._model;
+    SharedTreeModel m = _bldr._model;
```
Comment (mmalohlava) [inline h2o-algos/src/main/java/hex/tree/Score.java]: Side note (unrelated from the change): having model serialized is really bad, i still do not know why we need full model, not only selected info.

File: h2o-algos/src/main/java/hex/tree/Score.java [hunk 3]
```diff
@@ -40,7 +56,7 @@
     // If this is a score-on-train AND DRF, then oobColIdx makes sense,
     // otherwise this field is unused.
     final int oobColIdx = _bldr.idx_oobt();
-    _mb = m.makeMetricBuilder(domain);
+    _mb = (ModelMetricsSupervised.MetricBuilderSupervised) m.makeMetricBuilder(domain);
```
Comment (mmalohlava) [inline h2o-algos/src/main/java/hex/tree/Score.java]: Would it better to parameterize makeMetricBuilder by `<T>`? At least the cast would not be necessary.

File: h2o-core/src/main/java/hex/ModelMetricsBinomial.java [hunk 4]
```diff
@@ -167,37 +169,62 @@ static public ModelMetricsBinomial make(Vec targetClassProbs, Vec actualLabels,
...
       if (m!=null) m.addModelMetrics(mm);
       return mm;
     }
+
+    private GainsLift calculateGainsLift(Model m, Frame preds, Vec resp, Vec weights) {
+      GainsLift gl = null;
+      try {
+        gl = new GainsLift(preds.lastVec(), resp, weights);
+        gl.exec(m != null ? m._output._job : null);
+      } catch(Throwable t) { // TODO: Why do we need to catch Throwable here?
```
Comment (mmalohlava) [inline h2o-core/src/main/java/hex/ModelMetricsBinomial.java]: I would change it to Exception, at least we will not hide any Errors

File: h2o-algos/src/main/java/hex/tree/Score.java (hunk 1 above)
Comment (michalkurka) [reply h2o-algos/src/main/java/hex/tree/Score.java]: The frame is are actually not installed in DKV. For this case I think it makes a very little difference whether we serialize a key or the frame.

File: h2o-algos/src/main/java/hex/tree/Score.java (hunk 3 above)
Comment (michalkurka) [reply h2o-algos/src/main/java/hex/tree/Score.java]: Yes, this is a good point.

File: h2o-algos/src/main/java/hex/tree/Score.java (hunk 2 above)
Comment (michalkurka) [reply h2o-algos/src/main/java/hex/tree/Score.java]: This is a good point. We don't need the whole model object

**Comment by michalkurka** [comment]
@h2o-ops please test!

File: h2o-algos/src/main/java/hex/tree/Score.java (hunk 1 above)
Comment (mmalohlava) [reply h2o-algos/src/main/java/hex/tree/Score.java]: ok

**Pull request merged**